pynapple/process/filtering.py

```diff
@@ -0,0 +1,429 @@
+"""Filtering module."""
+
+import inspect
+from functools import wraps
+from numbers import Number
+
+import numpy as np
+from scipy.signal import butter, sosfiltfilt
+
+from .. import core as nap
+
+
+def _get_butter_coefficients(cutoff, filter_type, sampling_frequency, order=4):
+    return butter(order, cutoff, btype=filter_type, fs=sampling_frequency, output="sos")
+
+
+def _compute_butterworth_filter(
+    data, cutoff, sampling_frequency=None, filter_type="bandpass", order=4
+):
+    """
+    Apply a Butterworth filter to the provided signal.
+    """
+    sos = _get_butter_coefficients(cutoff, filter_type, sampling_frequency, order)
+    out = np.zeros_like(data.d)
+    for ep in data.time_support:
+        slc = data.get_slice(start=ep.start[0], end=ep.end[0])
+        out[slc] = sosfiltfilt(sos, data.d[slc], axis=0)
+
+    kwargs = dict(t=data.t, d=out, time_support=data.time_support)
+    if isinstance(data, nap.TsdFrame):
+        kwargs["columns"] = data.columns
+    return data.__class__(**kwargs)
+
+
+def _compute_spectral_inversion(kernel):
+    """
+    Compute the spectral inversion.
+    Parameters
+    ----------
+    kernel: ndarray
+
+    Returns
+    -------
+    ndarray
+    """
+    kernel *= -1.0
+    kernel[len(kernel) // 2] = 1.0 + kernel[len(kernel) // 2]
+    return kernel
+
+
+def _get_windowed_sinc_kernel(fc, filter_type="lowpass", transition_bandwidth=0.02):
+    """
+    Get the windowed-sinc kernel.
+    Smith, S. (2003). Digital signal processing: a practical guide for engineers and scientists.
+    Chapter 16, equation 16-4
+
+    Parameters
+    ----------
+    fc: float or tuple of float
+        Cutting frequency between 0 and 0.5. Single float for 'lowpass' and 'highpass'. Tuple of float for
+        'bandpass' and 'bandstop'.
+    filter_type: str
+        Either 'lowpass', 'highpass', 'bandstop' or 'bandpass'.
+    transition_bandwidth: float
+        Percentage between 0 and 0.5
+    Returns
+    -------
+    np.ndarray
+    """
+    M = int(np.rint(20.0 / transition_bandwidth))
+    x = np.arange(-(M // 2), 1 + (M // 2))
```

Comment:
```suggestion
    # This assures len(x) is odd, which is mandatory for spectral inversion to work.
    x = np.arange(-(M // 2), 1 + (M // 2))
```

I'll add a comment saying that you are making sure that you have an odd filter length because it is required for the filter to have the right property. I am worried that if we don't write it as a comment we will forget.